lib/object.c

```diff
@@ -338,8 +338,13 @@ int nl_object_identical(struct nl_object *a, struct nl_object *b)
 		req_attrs_b = req_attrs_a;
 	}
 
-	req_attrs_a &= a->ce_mask;
-	req_attrs_b &= b->ce_mask;
+	if (ops->oo_id_attrs_get || ops->oo_id_attrs) {
+		req_attrs_a |= a->ce_mask;
+		req_attrs_b |= b->ce_mask;
+	} else {
+		req_attrs_a &= a->ce_mask;
+		req_attrs_b &= b->ce_mask;
```

Comment:
This doesn't seem right to me. I think the previous code is correct.

`nl_object_identical()` compares objects not fully, but whether their primary key attribute correspond.

whether `ops` is determining the key attributes via `ops->oo_id_attrs_get`, `ops->oo_id_attrs` or as `UINT64_MAX`, doesn't really matter -- except that `ops->oo_id_attrs_get` allows to determine the key attributes based on the object-instance (and not statically for all objects of the given type). Not setting `ops->oo_id_attrs` is merely a convenience shortcut to set it to `UINT64_MAX`. It does not imply any different behavior during the comparison.

I think the current code is correct. We determine the key attributes `req_attrs_a` (per object instance). Afterwards, we filter out the unset attrbitues (`&= a->ce_mask`) and compare only the set ones. The point of this `&=` is that if the objects don't have all required-attributes set, that they still can compare as identical as long as all their set attributes agree.

Reply:
You are right. I was partially confused by the name of the function _nl_object_identical()_ which made me think that it should make exhaustive object comparison, and partially by our code that was built around previous (erroneous) behavior of given function. All that directed me towards misconception so thanks a lot for taking the time explain functions conceived behavior. Feel free to close this PR.